internal/goldendataset/span_generator.go

```diff
@@ -0,0 +1,497 @@
+	SpanStatusPermissionDenied:   otlptrace.Status_PermissionDenied,
+	SpanStatusResourceExhausted:  otlptrace.Status_ResourceExhausted,
+	SpanStatusFailedPrecondition: otlptrace.Status_FailedPrecondition,
+	SpanStatusAborted:            otlptrace.Status_Aborted,
+	SpanStatusOutOfRange:         otlptrace.Status_OutOfRange,
+	SpanStatusUnimplemented:      otlptrace.Status_Unimplemented,
+	SpanStatusInternalError:      otlptrace.Status_InternalError,
+	SpanStatusUnavailable:        otlptrace.Status_Unavailable,
+	SpanStatusDataLoss:           otlptrace.Status_DataLoss,
+	SpanStatusUnauthenticated:    otlptrace.Status_Unauthenticated,
+}
+
+var statusMsgMap = map[PICTInputStatus]string{
+	SpanStatusOk:                 "",
+	SpanStatusCancelled:          "Cancellation received",
+	SpanStatusUnknownError:       "",
+	SpanStatusInvalidArgument:    "parameter is required",
+	SpanStatusDeadlineExceeded:   "timed out after 30002 ms",
+	SpanStatusNotFound:           "/dragons/RomanianLonghorn not found",
+	SpanStatusAlreadyExists:      "/dragons/Drogon already exists",
+	SpanStatusPermissionDenied:   "tlannister does not have write permission",
+	SpanStatusResourceExhausted:  "ResourceExhausted",
+	SpanStatusFailedPrecondition: "33a64df551425fcc55e4d42a148795d9f25f89d4 has been edited",
+	SpanStatusAborted:            "",
+	SpanStatusOutOfRange:         "Range Not Satisfiable",
+	SpanStatusUnimplemented:      "Unimplemented",
+	SpanStatusInternalError:      "java.lang.NullPointerException",
+	SpanStatusUnavailable:        "RecommendationService is currently unavailable",
+	SpanStatusDataLoss:           "",
+	SpanStatusUnauthenticated:    "nstark is unknown user",
+}
+
+func GenerateSpans(count int, startPos int, pictFile string, random io.Reader) ([]*otlptrace.Span, int, error) {
+	pairsData, err := loadPictOutputFile(pictFile)
+	if err != nil {
+		return nil, 0, err
+	}
+	pairsTotal := len(pairsData)
+	spanList := make([]*otlptrace.Span, count)
+	index := startPos + 1
+	var inputs []string
+	var traceID []byte
+	var parentID []byte
+	for i := 0; i < count; i++ {
+		inputs = pairsData[index]
+		switch PICTInputParent(inputs[SpansColumnParent]) {
+		case SpanParentRoot:
+			traceID = generateTraceID(random)
+			parentID = nil
+		case SpanParentChild:
+			// use existing if available
+			if traceID == nil {
+				traceID = generateTraceID(random)
+			}
+			if parentID == nil {
+				parentID = generateSpanID(random)
+			}
+		}
+		spanName := generateSpanName(PICTInputAttributes(inputs[SpansColumnAttributes]), i)
+		spanList[i] = GenerateSpan(traceID, PICTInputTracestate(inputs[SpansColumnTracestate]), parentID, spanName,
+			PICTInputKind(inputs[SpansColumnKind]), PICTInputAttributes(inputs[SpansColumnAttributes]),
+			PICTInputSpanChild(inputs[SpansColumnEvents]), PICTInputSpanChild(inputs[SpansColumnLinks]),
+			PICTInputStatus(inputs[SpansColumnStatus]), random)
+		parentID = spanList[i].SpanId
+		index++
+		if index >= pairsTotal {
```

Comment:
Nit: the check probably should be at the start of the loop, not at the end, since in line 80 nothing prevents `index` from being more than len(pairsData).

Reply:
Fixed